src/cli/commands/add.js

```diff
@@ -55,23 +56,22 @@ export class Add extends Install {
    */
   getPatternVersion(pattern: string, pkg: Manifest): string {
     const {exact, tilde} = this.flags;
-    const parts = PackageRequest.normalizePattern(pattern);
+    const {hasVersion, range} = PackageRequest.normalizePattern(pattern);
     let version;
+
     if (getExoticResolver(pattern)) {
       // wasn't a name/range tuple so this is just a raw exotic pattern
       version = pattern;
-    } else if (parts.hasVersion && parts.range) {
+    } else if (hasVersion && range && (semver.satisfies(pkg.version, range) || getExoticResolver(range))) {
```

Comment:
Yeah, unfortunately `getExoticResolver` actually only uses `Resolver.isVersion` so if you pass `package@githubUser/packageRepo#master` for instance, it doesn't match. This was only visible after I made the fix.